platforms/jvm/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/junit/JUnitTestEventAdapter.java

```diff
@@ -210,162 +260,129 @@ private Object startParentIfNeeded(Description description) {
      * @param now the current time
      */
     private void startParentByNodeIfNeeded(TestNode parent, long now) {
-        Object grandparentId = startParentIfNeeded(parent.description);
-        synchronized (lock) {
-            if (activeParents.contains(parent.description)) {
-                return;
-            }
-            activeParents.addLast(parent.description);
-        }
-        String rootName = grandparentId == null ? this.rootName : className(parent.description);
-        if (rootName == null) {
-            throw new AssertionError("No class name found for " + parent.description);
-        }
-        TestDescriptorInternal parentDescriptor;
-        if (grandparentId != null && supportsTestClassMethod() && getTestClassIfPossible(parent.description) == null) {
-            // When Description.getTestClass() returns null and we have a grandparent,
-            // it indicates "suites" of parameterized tests, or some other kind of synthetic grouping.
-            // Avoid treating these as classes.
-            parentDescriptor = new DefaultTestSuiteDescriptor(parent.resolveId(), rootName);
-        } else {
-            parentDescriptor = new DefaultTestClassDescriptor(parent.resolveId(), rootName, classDisplayName(rootName));
+        Object grandparentId = startParentIfNeeded(parent);
+        if (executing.contains(parent)) {
+            return;
         }
-        resultProcessor.started(parentDescriptor, new TestStartEvent(now, grandparentId));
+        startNode(parent, new TestStartEvent(now, grandparentId));
+    }
+
+    private void startNode(TestNode node, TestStartEvent event) {
+        testsStarted = true;
+        executing.add(node);
+        executingStack.addLast(node);
+        getExecutingStackForCurrentThread().addLast(node);
+        resultProcessor.started(node.descriptor, event);
+        descriptionToStartedCount.compute(
+            node.description, (desc, count) -> count == null ? 1 : count + 1
+        );
+    }
+
+    private void completeNode(TestNode node, TestCompleteEvent event) {
+        resultProcessor.completed(node.descriptor.getId(), event);
+        executing.remove(node);
+        // The node is more likely to be at the end because we shouldn't finish parents before children
+        executingStack.removeLastOccurrence(node);
+        getExecutingStackForCurrentThread().removeLastOccurrence(node);
```

Comment:
Same here
```suggestion
        executing.remove(node);
        // The node is more likely to be at the end because we shouldn't finish parents before children
        executingStack.removeLastOccurrence(node);
        getExecutingStackForCurrentThread().removeLastOccurrence(node);
        resultProcessor.completed(node.descriptor.getId(), event);
```